### PR TITLE
Make RPC cancelable

### PR DESCRIPTION
**Pull request by pikatos**
Hi I'm working on a wallet and I need to cancel my calls. So there it is: cancelable RPCs. I should add some unit tests to it…

### CODE REVIEW

**Comment by fortmarek** [comment]
Great, I’ll take a look at it tomorrow, thanks!

File: TezosSwift/Client/Call.swift [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+extension Call {
+	public func flatMap<T>(_ transform: @escaping (Value) throws -> T) -> AnyCall<T.Value, T.Error> where T : Call, Error == T.Error {
+		return AnyCall<T.Value, T.Error> { completion in
+			let semaphore = DispatchSemaphore(value: 1)
+			var canceling = false
+			var cancelable: Cancelable?
+			cancelable = self.perform { result in
+				semaphore.wait(); defer { semaphore.signal() }
```
Comment (fortmarek) [inline TezosSwift/Client/Call.swift]: I don't really like having two statements on one line - do you have a specific reason to do so here?

File: TezosSwift/Client/Call.swift [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+						throw error
+					}
+				} catch let error as T.Error {
+					completion(.failure(error))
+				} catch {
+					completion(.failure(T.Error(error)))
+				}
+			}
+			return AnyCancelable {
+				semaphore.wait(); defer { semaphore.signal() }
```
Comment (fortmarek) [inline TezosSwift/Client/Call.swift]: Same comment regarding two statements on one line.

File: TezosSwift/Client/Call.swift [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+		if let error = error as? TezosError {
+			self = error
+			return
+		}
+		self = .unknown(message: error.localizedDescription)
+	}
+}
+
+extension Call {
+	public func flatMap<T>(_ transform: @escaping (Value) throws -> T) -> AnyCall<T.Value, T.Error> where T : Call, Error == T.Error {
```
Comment (fortmarek) [inline TezosSwift/Client/Call.swift]: Please, just delete the space between `T : Call`, ie `T: Call`

File: TezosSwift/Client/Call.swift [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+	}
+}
+
+extension TezosError: CustomErrorConvertible {
+	public init(_ error: Error) {
+		if let error = error as? TezosError {
+			self = error
+			return
+		}
+		self = .unknown(message: error.localizedDescription)
```
Comment (fortmarek) [inline TezosSwift/Client/Call.swift]: I'd prefer `if else` in this case to better state what the intent is.

File: TezosSwift/Client/Call.swift [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+public struct AnyCall<Value, Error: Swift.Error & CustomErrorConvertible>: Call {
+	
+	public typealias Body = (@escaping Completion) -> Cancelable?
+	
+	private let body: Body
+	
+	public init(_ body: @escaping Body) {
+		self.body = body
+	}
+	public init<T>(_ base: T) where T : Call, Value == T.Value, Error == T.Error {
```
Comment (fortmarek) [inline TezosSwift/Client/Call.swift]: Remove the space here, too.

File: TezosSwift/Client/TezosClient.swift [hunk 6]
```diff
@@ -545,7 +591,7 @@ public class TezosClient {
 
         if method == .post {
             do {
-                guard let payload = payload else { return }
+				guard let payload = payload else { return nil } // TODO: maybe call completion with .encryptionFailed
```
Comment (fortmarek) [inline TezosSwift/Client/TezosClient.swift]: What exactly do you mean by this TODO?

**Review by fortmarek — CHANGES_REQUESTED** [review]
Nice PR, I love the fact that this brings useful functionality while keeping the changes non-breaking.

I've requested a small few changes but I like the overall structure 👍

Before being able to merge, you should add at least some unit tests. It'd be great if you also added it into documentation and added a small example into the example project (if you do not feel up to these two, I can do it by myself, let me know)

File: TezosSwift/Client/Call.swift [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+	}
+}
+
+extension Call {
+	public func flatMap<T>(_ transform: @escaping (Value) throws -> T) -> AnyCall<T.Value, T.Error> where T : Call, Error == T.Error {
+		return AnyCall<T.Value, T.Error> { completion in
+			let semaphore = DispatchSemaphore(value: 1)
+			var canceling = false
+			var cancelable: Cancelable?
+			cancelable = self.perform { result in
```
Comment (fortmarek) [inline TezosSwift/Client/Call.swift]: I think it'd be better to make the `self` weak here, for leak-safety reasons.

**Comment by pikatos** [comment]
I'll try to work on it this week end. Thanks for your fast answer and your
work.

Le jeu. 14 févr. 2019 à 15:58, Marek Fořt <notifications@github.com> a
écrit :

File: TezosSwift/Client/TezosClient.swift (hunk 6 above)
Comment (pikatos) [reply TezosSwift/Client/TezosClient.swift]: In this case completion will never be called and could result in an unexpected behavior. User completion should always be called no matter what happens.

**Comment by pikatos** [comment]
Most of the integration tests fail on the master branch is it normal ?

**Comment by fortmarek** [comment]
Yes, failing integration tests are OK since they would have to batched to work (which I have not implemented as of now.)

**Comment by pikatos** [comment]
> Yes, failing integration tests are OK since they would have to batched to work (which I have not implemented as of now.)

**Comment by pikatos** [comment]
Oups I closed the pull request by mistake

**Comment by pikatos** [comment]
Could you check what I've done? I will add some comments to clarify what's what.

**Comment by fortmarek** [comment]
I will do it today's evening or tomorrow's morning, thanks!

File: TezosSwift/Client/TezosClient.swift (hunk 6 above)
Comment (fortmarek) [reply TezosSwift/Client/TezosClient.swift]: I agree, so could you please resolve this? I'd copy the `case decryptionFailed(reason: DecryptionReason)` from `TezosError` and implement it for encryption.

File: TezosSwift/Client/TezosError.swift [hunk 8]
```diff
@@ -89,6 +91,18 @@ public enum TezosError: Error {
     case parameterError(reason: ParameterReason)
     /// Error when decoding Tezos response data
     case decryptionFailed(reason: DecryptionReason)
-    ///Error when encrypting data 
+    /// Error when encrypting data
     case encryptionFailed(reason: Error)
+	/// User cancel
+	case userCancel
```
Comment (fortmarek) [inline TezosSwift/Client/TezosError.swift]: I'd maybe just go with `case canceled`? We could cancel the operation for other reasons than user canceling the operation, I think.

**Review by fortmarek — CHANGES_REQUESTED** [review]
I really like this, there's just two comments, otherwise I think it's fine.
And kudos for the tests 😎  They are quite meticulous.

**Comment by pikatos** [comment]
Thanks for your comments. Are you ok with the fix?

**Review by fortmarek — APPROVED** [review]

**Comment by fortmarek** [comment]
OK, I'm going to merge this now. I'll need to edit https://github.com/AckeeCZ/Tezos-iOS-Dev-Kit before marking this as a new release, I'll do it sometime this week. Thanks and let me know if there is anything else I can do for you!

**Pull request merged**